### PR TITLE
WIP: Added tabular explainer e2e testcase

### LINKED ISSUE #539: [Help Wanted] Add e2e tests for text/tabular/image explainers
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by janeman98**
**What this PR does / why we need it**:
Added tabular explainer e2e testcase. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #539 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

No change in image versions.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/736)

### CODE REVIEW

File: test/e2e/common/utils.py [hunk 1]
```diff
@@ -46,3 +46,24 @@ def predict(service_name, input_json):
         logging.info("Got response code %s, content %s", response.status_code, response.content)
         probs = json.loads(response.content.decode('utf-8'))["predictions"]
         return probs
+
+def explain(service_name, input_json):
+    isvc = KFServing.get(service_name, namespace=KFSERVING_TEST_NAMESPACE)
+    # temporary sleep until this is fixed https://github.com/kubeflow/kfserving/issues/604
+    time.sleep(10)
```
Comment (ellis-bigelow) [inline test/e2e/common/utils.py]: Is this enough? If there's a race, I'd like to overkill it so we never see it. 1 second okay?

File: test/e2e/common/utils.py [hunk 2]
```diff
@@ -46,3 +46,24 @@ def predict(service_name, input_json):
...
+    api_instance = client.CoreV1Api(client.ApiClient())
+    service = api_instance.read_namespaced_service("istio-ingressgateway", "istio-system", exact='true')
+    if service.status.load_balancer.ingress is None:
```
Comment (ellis-bigelow) [inline test/e2e/common/utils.py]: Can we generalize this to a helper to sahre between predict and explain?

File: test/e2e/explainer/test_tabular_explainer.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+api_version = constants.KFSERVING_GROUP + '/' + constants.KFSERVING_VERSION
+KFServing = KFServingClient(config_file="~/.kube/config")
+
+
+def test_tabular_explainer():
+    service_name = 'isvc-explainer-tabular'
+    default_endpoint_spec = V1alpha2EndpointSpec(
+        predictor=V1alpha2PredictorSpec(
+            min_replicas=1,
```
Comment (ellis-bigelow) [inline test/e2e/explainer/test_tabular_explainer.py]: This is the default, right?

File: test/e2e/explainer/test_tabular_explainer.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            sklearn=V1alpha2SKLearnSpec(
+                storage_uri='gs://seldon-models/sklearn/income/model',
+                resources=V1ResourceRequirements(
+                    requests={'cpu': '100m', 'memory': '1Gi'},
+                    limits={'cpu': '100m', 'memory': '1Gi'}))),
+        explainer=V1alpha2ExplainerSpec(
+            min_replicas=1,
```
Comment (ellis-bigelow) [inline test/e2e/explainer/test_tabular_explainer.py]: likewise

File: test/e2e/common/utils.py (hunk 1 above)
Comment (janeman98) [reply test/e2e/common/utils.py]: @ellis-bigelow 
Thanks for your review feedback. Please correct me if I have missed something here. The current code [sleep](https://docs.python.org/3/library/time.html?highlight=time#time.sleep) for 10 sec, are you saying to make it sleep longer, like 15 sec?

File: test/e2e/common/utils.py (hunk 2 above)
Comment (janeman98) [reply test/e2e/common/utils.py]: Sound good. Will do.

File: test/e2e/explainer/test_tabular_explainer.py (hunk 3 above)
Comment (janeman98) [reply test/e2e/explainer/test_tabular_explainer.py]: Agree. Changed in[ PR655](https://github.com/kubeflow/kfserving/pull/655)

File: test/e2e/explainer/test_tabular_explainer.py (hunk 4 above)
Comment (janeman98) [reply test/e2e/explainer/test_tabular_explainer.py]: Agree. Changed in [PR655](https://github.com/kubeflow/kfserving/pull/655)

File: test/e2e/common/utils.py (hunk 1 above)
Comment (ellis-bigelow) [reply test/e2e/common/utils.py]: For some reason I thought this was milliseconds. You're all over it!

File: test/e2e/common/utils.py [hunk 5]
```diff
@@ -30,12 +30,7 @@ def predict(service_name, input_json):
     isvc = KFServing.get(service_name, namespace=KFSERVING_TEST_NAMESPACE)
     # temporary sleep until this is fixed https://github.com/kubeflow/kfserving/issues/604
     time.sleep(10)
-    api_instance = client.CoreV1Api(client.ApiClient())
-    service = api_instance.read_namespaced_service("istio-ingressgateway", "istio-system", exact='true')
-    if service.status.load_balancer.ingress is None:
-        cluster_ip = service.spec.cluster_ip
-    else:    
-        cluster_ip = service.status.load_balancer.ingress[0].ip
+    cluster_ip = getCluster_ip()
```
Comment (ellis-bigelow) [inline test/e2e/common/utils.py]: nit: odd casing. How about get_cluster_ip()

**Comment by ellis-bigelow** [comment]
Nice work!
/approve

Feel free to get lgtm from anyone if I'm slow.

File: test/e2e/common/utils.py (hunk 5 above)
Comment (janeman98) [reply test/e2e/common/utils.py]: Will change

**Comment by yuzisun** [comment]
```
  [E 200418 07:33:40 web:1792] Uncaught exception POST /v1/models/isvc-explainer-tabular:explain (127.0.0.1)
       HTTPServerRequest(protocol='http', host='isvc-explainer-tabular-explainer-default.kfserving-ci-e2e-test.svc.cluster.local', method='POST', uri='/v1/models/
isvc-explainer-tabular:explain', version='HTTP/1.1', remote_ip='127.0.0.1')
       Traceback (most recent call last):
         File "/usr/local/lib/python3.7/site-packages/tornado/web.py", line 1701, in _execute
           result = method(*self.path_args, **self.path_kwargs)
         File "/kfserving/kfserving/handlers/http.py", line 61, in post
           response = model.explain(request)
         File "/alibiexplainer/alibiexplainer/explainer.py", line 74, in explain
           explanation = self.wrapper.explain(request["instances"])
         File "/alibiexplainer/alibiexplainer/anchor_tabular.py", line 89, in explain
           anchor_exp = self.anchors_tabular.explain(arr[0], **self.kwargs)
         File "/usr/local/lib/python3.7/site-packages/alibi/explainers/anchor_tabular.py", line 821, in explain
           for sampler in self.samplers:
       AttributeError: 'AnchorTabular' object has no attribute 'samplers'
   [E 200418 07:33:40 web:2250] 500 POST /v1/models/isvc-explainer-tabular:explain (127.0.0.1) 705.21ms
```
@janeman98 seems like it is failing here. @cliveseldon any idea?

**Comment by janeman98** [comment]
Thanks @yuzisun !

@cliveseldon 
There is a big change in /alibi/explainers/anchor_tabular.py in https://github.com/SeldonIO/alibi/pull/173. Will that relate to the error that Dan mentiioned above?

**Comment by cliveseldon** [comment]
> Thanks @yuzisun !
> 
> @cliveseldon
> There is a big change in /alibi/explainers/anchor_tabular.py in [SeldonIO/alibi#173](https://github.com/SeldonIO/alibi/pull/173). Will that relate to the error that Dan mentiioned above?

Yes the latest version of Alibi has some breaking changes but we access Alibi via the wrapper. So yes if the image is updated we should really fix at 0.3.0

https://github.com/kubeflow/kfserving/blob/3b282458eafe7fa423f7f4dcc34bd8f492c07884/python/alibiexplainer/setup.py#L36

I was hoping to update the alibi wrapper to work with latest.

**Comment by janeman98** [comment]
Thanks @cliveseldon 
Is there any existing issue opened for this? Or you want me to open one to keep track?

**Comment by cliveseldon** [comment]
Thanks @janeman98 have opened https://github.com/kubeflow/kfserving/issues/792

**Comment by yuzisun** [comment]
@janeman98 Can you help rebase #803 from master and see if this fixes the e2e test issue.

**Comment by janeman98** [comment]
> @janeman98 Can you help rebase #803 from master and see if this fixes the e2e test issue.

Thanks @yuzisun ! 
I have lost my development machine. Will resume after I have everything set up. I mark this PR **WIP** for now.

**Comment by yuzisun** [comment]
Sorry @janeman98 I am closing this in favor of https://github.com/kubeflow/kfserving/pull/865 since we need the test to release 0.4, but thanks for your awesome work!

**Comment by ns-jman** [comment]
@yuzisun 
np. Sorry for delay.

**Comment by janeman98** [comment]
@yuzisun 
Np. Sorry for delay.

**Pull request closed without merging**